command/agent.go

```diff
@@ -558,11 +570,31 @@ func (c *AgentCommand) Run(args []string) int {
 			Namespace:     namespace,
 			ExitAfterAuth: exitAfterAuth,
 		})
-		tsDoneCh = ts.DoneCh
 
-		go ah.Run(ctx, method, errCh)
-		go ss.Run(ctx, ah.OutputCh, sinks, errCh)
-		go ts.Run(ctx, ah.TemplateTokenCh, config.Templates, errCh)
+		g.Add(func() error {
+			return ah.Run(ctx, method)
+		}, func(error) {
+			cancelFunc()
+		})
+
+		g.Add(func() error {
+			err := ss.Run(ctx, ah.OutputCh, sinks)
+
+			// Wait until templates are rendered
+			<-ts.DoneCh
```

Comment:
Now that `ts.DoneCh` is closed in the `ts.Stop()` function i'm not sure this is truly waiting for `ts` to be done rendering? maybe you should go back to having the DoneCh closed on defer of the server?

Reply:
`ts.Stop()` is triggered by the interrupt func, which only happens once the actor exits. That should only happen if the render is finished (or if the render errors).

Reply:
> which only happens once the actor exits.

I don't think that's true. oklog/run runs the interrupt func if any of the actors exit.

Reply:
Though, perhaps we don't really care if the templates are rendered if there is another error. We are doing that because the template server can return after a single render, right?

Reply:
Right, but on the case where `exit_after_auth = true`, only `ss.Run`'s or `ts.Run`'s actor can exit right away, and we block on the former in here. The latter only exits after a successful render which then triggers its interrupt func, calling `ts.Stop()`, and unblocking this actor (`ss.Run`).

Reply:
Actually this is not quite right. I think this will deadlock since `ts.Run` is blocking on the return until the context is canceled, which won't happen since we're blocking the actor from returning here (and auth handler's actor won't return since it's not aware of this bool).

Reply:
This was causing `TestAgent_ExitAfterAuth` to fail, but should now be addressed by 4d73df2. Note that this makes `ss` and `ts` pretty tightly coupled so there's room for improvement, but it should get us through this hoop for the moment.